File: filter-pattern/js/src/AndCriteria.js
import { Criteria } from './Criteria.js'

// 定义And过滤标准
export class AndCriteria extends Criteria {
  constructor(criteria, otherCriteria) {
    super()
    this.criteria = criteria
    this.otherCriteria = otherCriteria
  }

  filter(persons) {
    const firstCriteriaPersons = this.criteria.filter(persons)
    return this.otherCriteria.filter(firstCriteriaPersons)
  }
}

File: filter-pattern/java/src/AndCriteria.java
package src;

import java.util.*;

// 定义And过滤标准
public class AndCriteria implements Criteria {

   private Criteria criteria;
   private Criteria otherCriteria;

   public AndCriteria(Criteria criteria, Criteria otherCriteria) {
      this.criteria = criteria;
      this.otherCriteria = otherCriteria;
   }

   @Override
   public List<Person> filter(List<Person> persons) {
      List<Person> firstCriteriaPersons = criteria.filter(persons);
      return otherCriteria.filter(firstCriteriaPersons);
   }
}